packages/server/test/notifier.test.ts

```diff
@@ -459,14 +468,171 @@ describe('Notifier', () => {
-      expect(mockSendEmail).toHaveBeenCalledTimes(1)
+      expect(mockEmailProvider.sentEmails.length).toBe(1)
+    })
+  })
+
+  describe('retries', () => {
+    it('should retry failed notifications', async () => {
+      // -------------------
+      // Initialize database
+      // -------------------
+
+      const alice = await createMember(1, 'alice', [{ kind: 'FORUM_POST_MENTION' }])
+
+      // -------------------
+      // Mock QN responses
+      // -------------------
+
+      mockRequest
+        .mockReturnValueOnce({
+          events: [
+            postAddedEvent(1, {
+              thread: 'foo',
+              threadAuthor: 'foo',
+              text: `Hi [@Alice](#mention?member-id=${alice.id})`,
+            }),
+          ],
+        })
+        .mockReturnValue({
+          events: [],
+        })
+
+      // Enable email provider fails
+      mockEmailProvider.shouldFail = true
+
+      // Run `run()` MAX_EMAIL_RETRY_COUNT times
+
+      for (let i = 0; i < EMAIL_MAX_RETRY_COUNT; i++) {
+        await run()
+
+        // -------------------
+        // Check notifications
+        // -------------------
+
+        const notifications = await prisma.notification.findMany()
+
+        expect(notifications).toContainEqual(
+          expect.objectContaining({
+            eventId: 'event:1',
+            memberId: alice.id,
+            kind: 'FORUM_POST_MENTION',
+            status: 'PENDING',
+            retryCount: i + 1,
+          })
+        )
+        expect(notifications).toHaveLength(1)
+      }
+
+      // Expect the notification to be marked as failed after retries are exhausted
+
+      await run()
+
+      const notifications = await prisma.notification.findMany()
+
+      expect(notifications).toContainEqual(
+        expect.objectContaining({
+          eventId: 'event:1',
+          memberId: alice.id,
+          kind: 'FORUM_POST_MENTION',
+          status: 'FAILED',
+          retryCount: EMAIL_MAX_RETRY_COUNT,
+        })
+      )
+
+      // Make sure no emails were sent
+
+      expect(mockEmailProvider.sentEmails.length).toBe(0)
+    })
+
+    it('should properly send retried notifications', async () => {
+      // -------------------
+      // Initialize database
+      // -------------------
+
+      const alice = await createMember(1, 'alice', [{ kind: 'FORUM_POST_MENTION' }])
+
+      // -------------------
+      // Mock QN responses
+      // -------------------
+
+      mockRequest
+        .mockReturnValueOnce({
+          events: [
+            postAddedEvent(1, {
+              thread: 'foo',
+              threadAuthor: 'foo',
+              text: `Hi [@Alice](#mention?member-id=${alice.id})`,
+            }),
+          ],
+        })
+        .mockReturnValue({
+          events: [],
+          forumPostByUniqueInput: {
+            author: { handle: 'author:handle' },
+            thread: { id: 'thread:id', title: 'thread:title' },
+            text: 'Lorem Ipsum',
+          },
```

Comment:
Is this a mistake ?
```suggestion
```

Reply:
Not sure what you mean, but that data is needed to create the notification for `postAddedEvent`

Reply:
Right! It's been a while... I forgot this mocks the result of all the gql requests